code_search/code-search.ipynb

```diff
@@ -0,0 +1,451 @@
+    "        --runner DataflowRunner \\\n",
+    "        --project \"${PROJECT}\" \\\n",
+    "        --target_dataset \"${TARGET_DATASET}\" \\\n",
+    "        --data_dir \"${WORKING_DIR}/data\" \\\n",
+    "        --job_name \"preprocess-github-dataset\" \\\n",
+    "        --temp_location \"${WORKING_DIR}/data/dataflow/temp\" \\\n",
+    "        --staging_location \"${WORKING_DIR}/data/dataflow/staging\" \\\n",
+    "        --worker_machine_type \"${WORKER_MACHINE_TYPE}\" \\\n",
+    "        --num_workers \"${NUM_WORKERS}\""
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "When completed successfully, this should create a dataset in `BigQuery` named `target_dataset`. Additionally, it also dumps CSV files into `data_dir` which contain training samples (pairs of function and docstrings) for our Tensorflow Model."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Prepare Dataset for Training\n",
+    "\n",
+    "In this step we will use `t2t-datagen` to convert the transformed data above into the `TFRecord` format. We will run this job on the Kubeflow cluster."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "%%bash\n",
+    "\n",
+    "ks apply code-search -c t2t-code-search-datagen"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Execute Tensorflow Training"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "%%bash\n",
+    "\n",
+    "ks apply code-search -c t2t-code-search-trainer"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Export Tensorflow Model"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "%%bash\n",
+    "\n",
+    "ks apply code-search -c t2t-code-search-exporter"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Compute Function Embeddings\n",
+    "\n",
+    "In this step, we will use the exported model above to compute function embeddings via another `Dataflow` pipeline. A `Python 2` module `code_search.dataflow.cli.create_function_embeddings` has been provided for this purpose. A list of all possible arguments can be seen below."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "%%bash\n",
+    "\n",
+    "python2 -m code_search.dataflow.cli.create_function_embeddings -h"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Configuration\n",
+    "\n",
+    "First, select a Exported Model version from the `${WORKING_DIR}/output/export/Servo`. This should be name of a folder with UNIX Seconds Timestamp like `1533685294`."
```

Comment:
Can we write some python code to automatically do this?

Reply:
The `gsutil` CLI should be easier though passing back the variables from bash to python might not be possible. I will check this or fallback to python.

Reply:
You can use this python client library with GCS
https://googlecloudplatform.github.io/google-cloud-python/latest/